src/main/java/org/opentripplanner/apis/gtfs/PatternByServiceDatesFilter.java

```diff
@@ -0,0 +1,100 @@
+package org.opentripplanner.apis.gtfs;
+
+import java.time.LocalDate;
+import java.util.Collection;
+import java.util.Objects;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+import org.opentripplanner.apis.gtfs.generated.GraphQLTypes.GraphQLServiceDateFilterInput;
+import org.opentripplanner.transit.model.network.Route;
+import org.opentripplanner.transit.model.network.TripPattern;
+import org.opentripplanner.transit.service.TransitService;
+
+/**
+ * Encapsulates the logic to filter patterns by the service dates that they operate on. It also
+ * has a method to filter routes by checking if their patterns operate on the required days
+ */
+public class PatternByServiceDatesFilter {
+
+  private final TransitService transitService;
+  private final LocalDate startInclusive;
+  private final LocalDate endInclusive;
+
+  /**
+   *
+   * @param transitService
+   * @param startInclusive The inclusive start date to check the patterns for. If null then no start
+   *                       date is defined and this will therefore match all dates.
+   * @param endInclusive The inclusive end date to check the patterns for. If null then no end date
+   *                     is defined this will therefore match all dates.
+   */
+  PatternByServiceDatesFilter(
+    TransitService transitService,
+    @Nullable LocalDate startInclusive,
+    @Nullable LocalDate endInclusive
+  ) {
+    Objects.requireNonNull(transitService);
+    this.transitService = transitService;
+
+    // optional
+    this.startInclusive = startInclusive;
+    this.endInclusive = endInclusive;
+
+    if (startInclusive != null && endInclusive != null && startInclusive.isAfter(endInclusive)) {
+      throw new IllegalArgumentException("start must be before end");
+    }
```

Comment:
Should we also validate that either start or end is defined, otherwise does using this filter make sense?

Reply:
This check actually happens here: https://github.com/opentripplanner/OpenTripPlanner/blob/e6971b4d786e3c94b02e52aca1512c81079ac887/src/main/java/org/opentripplanner/apis/gtfs/datafetchers/RouteImpl.java#L184

However, it wouldn't be a problem to also check in the constructor to make it extra sure that no-one misuses this class in the future. Do you think that's a good idea?

Reply:
I don't have a strong preference. I don't know if this will be reused anywhere else.

Reply:
https://github.com/opentripplanner/OpenTripPlanner/pull/5869/commits/3c3bd18df7264042cb5059c2845b98013a714ecb